trace/trace.go

```diff
@@ -123,73 +107,83 @@ func WithSpan(parent context.Context, s *Span) context.Context {
 	// If RecordEvents is false, this option has no effect.
 	RegisterNameForLocalSpanStore bool
+
+	// Tags are the tags applied to the span. Optional.
+	Tags *tag.Map
 }
 
 // StartSpan starts a new child span of the current span in the context.
 //
 // If there is no span in the context, creates a new trace and span.
-func StartSpan(ctx context.Context, name string) context.Context {
+func StartSpan(ctx context.Context, name string) (context.Context, *Span) {
 	parentSpan, _ := ctx.Value(contextKey{}).(*Span)
-	return WithSpan(ctx, parentSpan.StartSpanWithOptions(name, StartSpanOptions{}))
+	span := NewSpan(name, parentSpan, StartOptions{})
+	return WithSpan(ctx, span), span
 }
 
 // StartSpanWithOptions starts a new child span of the current span in the context.
 //
 // If there is no span in the context, creates a new trace and span.
-func StartSpanWithOptions(ctx context.Context, name string, o StartSpanOptions) context.Context {
+func StartSpanWithOptions(ctx context.Context, name string, o StartOptions) (context.Context, *Span) {
 	parentSpan, _ := ctx.Value(contextKey{}).(*Span)
-	return WithSpan(ctx, parentSpan.StartSpanWithOptions(name, o))
+	span := NewSpan(name, parentSpan, o)
+	return WithSpan(ctx, span), span
 }
 
-// StartSpanWithRemoteParent starts a new child span with the given parent SpanContext.
-//
-// If there is an existing span in ctx, it is ignored -- the returned Span is a
-// child of the span specified by parent.
-func StartSpanWithRemoteParent(ctx context.Context, name string, parent SpanContext, o StartSpanOptions) context.Context {
-	return WithSpan(ctx, NewSpanWithRemoteParent(name, parent, o))
-}
-
-// StartSpan starts a new child span.
+// EndSpan ends the current span.
 //
-// If s is nil, creates a new trace and span, like the function NewSpan.
-func (s *Span) StartSpan(name string) *Span {
-	return s.StartSpanWithOptions(name, StartSpanOptions{})
+// The context passed to EndSpan will still refer to the now-ended span, so any
+// code that adds more information to it, like SetSpanStatus, should be called
+// before EndSpan.
+func EndSpan(ctx context.Context) {
+	s, ok := ctx.Value(contextKey{}).(*Span)
+	if !ok {
+		return
+	}
+	s.End(EndOptions{Tags: tag.FromContext(ctx)})
 }
 
-// StartSpanWithOptions starts a new child span using the given options.
+// StartSpanWithRemoteParent starts a new child span with the given parent SpanContext.
 //
-// If s is nil, creates a new trace and span, like the function NewSpan.
-func (s *Span) StartSpanWithOptions(name string, o StartSpanOptions) *Span {
-	if s != nil {
-		return startSpanInternal(name, true, s.spanContext, false, o)
-	}
-	return startSpanInternal(name, false, SpanContext{}, false, o)
+// If there is an existing span in ctx, it is ignored -- the returned Span is a
+// child of the span specified by parent.
+func StartSpanWithRemoteParent(ctx context.Context, name string, parent SpanContext, o StartOptions) (context.Context, *Span) {
```

Comment:
I don't think this convenience method is carrying it's weight.

Reply:
Agreed.